xxxx-ewallet-blockchain-integration.md

```diff
@@ -0,0 +1,313 @@
+7. The user then needs to request a quote from the DEX directly in order to know how many of his choice of currency are needed to send 10 `ABC` to eWallet A's user. An event is sent to eWallet A/B about the progress.
+8. The user receives the quote back from the DEX on the Mobile application and approves it. An event is sent to eWallet A/B about the progress.
+9. The user deposits funds on Plasma in his choice of currency. This should be done beforehand unless the blockchain can handle fast-enough transaction.
+10. The user places the exchange order to get 10 `ABC` from the mobile application. Once the order has been matched, the user needs to send the funds (10 `ABC`) to eWallet A address on Plasma. An event is sent to eWallet A about the progress.
+11. eWallet A receives the Plasma event through the watcher, confirm the consumption and notifies the user. eWallet B forward the event to the user's mobile application to let him know the transaction has been confirmed.
+12. eWallet A's user receives an event in the websocket channel for the transaction request.
+
+### Implementation
+
+#### Code organization
+
+Two new sub applications will be added to the eWallet project: the `blockchain_interface` and the `omg_adapter`.
+
+##### `blockchain_interface`
+
+The blockchain interface will define functions that need to be implemented in the adapters.
+
+- `create_account`
+- `create_token`
+- `mint`
+- `burn`
+- `send_chain_transaction`
+- `send_offchain_transaction`
+- `get_quote`
+- `place_exchange_order`
+- `register_listener`
+
+##### `omg_adapter`
+
+- `create_account` (creates an Ethereum account)
+- `create_eth_token` (publish a token smart contract on Ethereum)
+- `mint` (adds more value to an existing token on Ethereum)
+- `deposit` (deposit funds to the specified Plasma contract)
+- `exit` (exit funds from the specified Plasma contract)
+- `send_chain_transaction` (sends an Ethereum transaction)
+- `send_offchain_transaction` (sends a Plasma transaction)
+- `get_quote` (get a quote from Exchange)
+- `place_exchange_order` (place an order for exchange)
+- `register_listener` (register a GenServer as a listener to events from Ethereum and Plasma)
+
+#### Database Changes
+
+Table name: `contract_templates`
+
+- `uuid`
+- `name`
+- `content`
+- `inserted_at`
+- `updated_at`
+
+Table name: `blockchain_wallet`
+
+- `uuid`
+- `address`
+- `pub_key`
+- `encrypted_priv_key` (only for `hot` wallets, encrypted using a key defined in ENV variables)
+- `type` (`hot`|`cold`)
+- `primary` (`true`|`false`)
+- `inserted_at`
+- `updated_at`
+
+Table name: `limits`
+
+- `uuid`
+- `token_id`
+- `low_amount`
+- `high_amount`
+- `limit_type` (`percent`|`value`)
+- `blockchain_wallet_uuid`
+- `inserted_at`
+- `updated_at`
+
+In case a blockchain wallet goes below or above the limits, a user will be notified that an action is required to keep the system secure (like moving funds in/out from/to a cold wallet.
```

Comment:
Nice! I think this eventually could be extended into a one-way [sweep account](https://en.wikipedia.org/wiki/Sweep_account) where if a hot wallet goes beyond `high_amount`, it triggers an auto-transfer of the extra amount to a cold wallet.